### ISSUE TITLE
Rule Proposal: JSX Fragment shorthand

### ISSUE THREAD

**Issue by dkrutsko** [yannickcr/eslint-plugin-react#1661]
Hey there, apologies if this was brought up in the past but I'd like to propose a new rule for the [ReactFragment](https://reactjs.org/docs/fragments.html) feature in React 16. Specifically, the rule would allow you to select between either the [shorthand](https://reactjs.org/docs/fragments.html#short-syntax) version `<>` or the `<React.Fragment>` version.

**Comment by ljharb** [comment]
This should only apply when the React pragma is >= 16.2; on earlier versions, it should probably warn on both.

**Comment by alexzherdev** [comment]
This also depends on Babel 7, right?

**Comment by dkrutsko** [comment]
@alexzherdev I believe you're right, I've recently updated to Babel 7 and I remember it coming up in the patch notes.

**Comment by alexzherdev** [comment]
So if we have a setting for the name of the fragment tag (i.e. `Fragment` vs `React.Fragment`), then this rule could potentially even be autofixable? Otherwise it wouldn't be safe to fix `<Fragment>` to `<>`; and in the reverse direction we wouldn't know whether to fix `<>` to be `<React.Fragment>` or `<Fragment>`.

**Comment by ljharb** [comment]
@alexzherdev yes, I'd expect that just like we have a pragma for React and createElement, we'd need one for Fragment.

Autofixing to `<>` is thus safe; i think we'd just always autofix to `React.Fragment` since `React` is sure to be in scope.

**Comment by alexzherdev** [comment]
Hmm, ok. I’m still trying to understand. I was thinking of the pragma as a way for the user to customize their preferred usage of
```jsx
<React.Fragment>
```
vs
```jsx
import { Fragment } from 'react';
<Fragment>
```
Setting it to `Fragment` would mean that it’s safe to assume `<Fragment>` to refer to React Fragment, not some user-defined component.
If it’s always equal to just `Fragment`, how would we handle the second form in the “always” case? Or would we just ignore it?

**Comment by alexzherdev** [comment]
Of course - that was a simplified example, I fixed it. My point is that there’s at least two ways you could refer to `<React.Fragment>`.

**Comment by ljharb** [comment]
Oh sure - i don’t think we have to account for that in the first iteration, tho.

**Comment by animanoir** [comment]
Thanks for this. Now how do I disabled it?

**Comment by ljharb** [comment]
@animanoir the same way you disable any rule; with eslint config, or an override comment.

**Comment by animanoir** [comment]
just in case someones search here :

`    "react/jsx-fragments": ["off"]`
